### PR TITLE
Updated for phpunit 9 and php8

### LINKED ISSUE #11: PHP 8.0 support
### Feature Request

|    Q        |   A
|------------ | ------
| New Feature | yes

#### Summary

To be prepared for the december release of PHP 8.0, this repository has some additional TODOs to be tested against the new major version.

In order to make this repository compatible, one has to follow these steps:

- [ ] Modify `composer.json` to provide support for PHP 8.0 by adding the constraint `~8.0.0`
- [ ] Modify `composer.json` to drop support for PHP less than 7.3
- [ ] Modify `composer.json` to implement phpunit 9.3 which supports PHP 7.3+
- [ ] Modify `.travis.yml` to ignore platform requirements when installing composer dependencies (simply add `--ignore-platform-reqs` to `COMPOSER_ARGS` env variable)
- [ ] Modify `.travis.yml` to add PHP 8.0 to the matrix (*NOTE*: Do not allow failures as PHP 8.0 has a feature freeze since 2020-08-04!)
- [ ] Modify source code in case there are incompatibilities with PHP 8.0

### PR DESCRIPTION
**Pull request by far-blue**
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | yes

### Description

Fixes issue #11 - PHP 8.0 support.

* Adds php 8.0 as a supported php version in composer.json.
* Updates phpunit to v9.5 in composer.json.
* Fixes changes to phpunit function definitions in subclasses.
* Replaces deprecated assertAttribute* methods by using subclasses that provide public getters for protected attributes.
* Replaces assertContains() method uses with assertStringContainsString() calls as per phpunit 9 changes.
* Attempted update of travis config.

I have made a best guess with travis as I neither use it or can test it.

Also note that the version of php_codesniffer specified in the laminas-coding-standard package won't run on php 8.0. I tested coding standard compliance using php 7.3.

Please let me know if I've missed anything! :)

### CODE REVIEW

**Comment by Ocramius** [comment]
Since travis-ci is dead, I cannot validate this patch: could you copy over https://github.com/laminas/laminas-crypt/blob/a058eeb2fe57824b958ac56753faff790a649e18/.github/workflows/continuous-integration.yml to validate this patch, please?

Note: not your fault - travis-ci is simply dead due to corporate greed :shrug:

**Comment by far-blue** [comment]
@Ocramius done :)

File: composer.json
Comment (froschdesign) [inline composer.json]: PHP does not follow [semantic versioning](https://semver.org).
```suggestion
        "php": "^7.3 || ~8.0.0",
```

File: composer.json
Comment (froschdesign) [inline composer.json]: Version 2 of laminas-servicemanager will no longer work with the newer PHP versions.

File: composer.json
Comment (far-blue) [reply composer.json]: That's a decision for the laminas maintainers really. I've just replicated the existing style used by them on master.

Comment (froschdesign) [reply composer.json]: See the related issue report: #11 

> Modify composer.json to provide support for PHP 8.0 by adding the constraint `~8.0.0`

File: composer.json
Comment (far-blue) [reply composer.json]: I believe composer will handle this and use v3 with php8. Version 2 of service manager should still work with php7.3 and php7.4. Or have I missed something?

Comment (froschdesign) [reply composer.json]: Sorry, yes, it works, but the old version can be dropped here.

File: composer.json
Comment (far-blue) [reply composer.json]: Ah, sorry, yeah, missed that :)

File: composer.json
Comment (far-blue) [reply composer.json]: if I drop it, won't it potentially break code in projects using this package unnecessarily?

Comment (froschdesign) [reply composer.json]: Version 3 is still there and if someone explicitly relies on version 2, the newer version of this package would not be installed.

Comment (far-blue) [reply composer.json]: I feel this change is beyond the original scope so unless it is going to prevent acceptance of the PR I'd prefer to leave the change to another PR.

**Comment by far-blue** [comment]
Any chance of this PR being accepted and merged any time?

**Comment by jmontoyaa** [comment]
The checks passed, It will be great if it can be added !

File: .travis.yml
Comment (PowerKiKi) [inline .travis.yml]: The entire file `.travis.yml` should be deleted, since we start to use GitHub Actions with this PR.

File: test/Assertion/TestSubclasses/AssertionAggregate.php
Comment (PowerKiKi) [inline test/Assertion/TestSubclasses/AssertionAggregate.php]: According to https://docs.laminas.dev/laminas-coding-standard/v2/coding-style-guide/#9-commenting-and-docblocks, unnecessary comments should not be introduced.

```suggestion
```

File: test/Assertion/TestSubclasses/AssertionAggregate.php
Comment (PowerKiKi) [inline test/Assertion/TestSubclasses/AssertionAggregate.php]: Idem

```suggestion
```

File: test/Assertion/TestSubclasses/CallbackAssertion.php
Comment (PowerKiKi) [inline test/Assertion/TestSubclasses/CallbackAssertion.php]: Idem, and comment is incorrect anyway

```suggestion
```

**Review by PowerKiKi — CHANGES_REQUESTED** [review]
There are a few minor improvements that should be made before merging.

**Comment by far-blue** [comment]
Eugh, when you 'commit suggestion' from GitHub it then failed DCO *sigh*

**Review by PowerKiKi — APPROVED** [review]

**Comment by jmontoyaa** [comment]
Merge ? :innocent:

**Comment by far-blue** [comment]
Is there anything still preventing this PR being accepted and merged?

File: composer.json
Comment (froschdesign) [reply composer.json]: @far-blue
Please remove version 2 of laminas-servicemanager, it cannot be installed with the specified PHP versions.

Comment (far-blue) [reply composer.json]: I feel this change is beyond the original scope so unless it is going to prevent acceptance of the PR I'd prefer to leave the change to another PR.

**Comment by far-blue** [comment]
Other than a requested change to composer.json that I consider out of scope and better suited to a separate PR because it introduces a breaking change for people using the package on older php versions, is there anything else preventing this PR being accepted and merged?

**Review by froschdesign — CHANGES_REQUESTED** [review]
@far-blue 
Please remove version 2 of laminas-servicemanager, it cannot be installed with the specified PHP versions.

> because it introduces a breaking change for people using the package on older php versions

The new version of this package can not be installed with older versions of PHP. So no breaking changes here.

**Comment by far-blue** [comment]
> @far-blue
> Please remove version 2 of laminas-servicemanager, it cannot be installed with the specified PHP versions.

It can still be used with php 7.3 and php 7.4 and this package does support php 7.3 and php 7.4. Composer will do it's job and correctly handle installing whichever version of laminas-servicemanager is appropriate.

**Comment by froschdesign** [comment]
@far-blue 
> It can still be used with php 7.3 and php 7.4 and this package does support php 7.3 and php 7.4. Composer will do it's job and correctly handle installing whichever version of laminas-servicemanager is appropriate.

laminas-servicemanager 2.7 does not support PHP 7.3 or PHP 7.4.

**Comment by far-blue** [comment]
> @far-blue
> 
> > It can still be used with php 7.3 and php 7.4 and this package does support php 7.3 and php 7.4. Composer will do it's job and correctly handle installing whichever version of laminas-servicemanager is appropriate.
> 
> laminas-servicemanager 2.7 does not supports PHP 7.3 or PHP 7.4.

That's not what you said previously - so which is it?

**Comment by far-blue** [comment]
packages reports laminas-sevrvicemanager v2.7.11 requires `php: ^5.5 || ^7.0`

**Comment by froschdesign** [comment]
@far-blue 
> packages reports laminas-sevrvicemanager v2.7.11 requires `php: ^5.5 || ^7.0`

Therefore, the version 2.7 can be removed without any problems.

**Comment by far-blue** [comment]
> @far-blue
> 
> > packages reports laminas-sevrvicemanager v2.7.11 requires `php: ^5.5 || ^7.0`
> 
> Therefore, the version 2.7 can be removed without any problems.

Service Manager supports php v7.0 through 7.4. This PR allows permissions-acl to support php v7.3 through 7.4 and then 8.0.x. That leaves php 7.3.x and php 7.4.x that are supported by both. It's not a requirement of php 8 support (i.e. the focus of this PR) to remove service manager v2 support. I consider it a disservice to other developers to arbitrarily remove support where compatibility exists, especially at no effort.

Do as you wish with the PR - I'm as busy as the next developer and switched over to my own fork of this package weeks ago in order to remove the block on support for php 8.

**Comment by jmontoyaa** [comment]
@far-blue I also want to use your fork, are you using this one?

https://github.com/far-blue/laminas-permissions-acl/tree/feature/php8.0_support

**Comment by far-blue** [comment]
> @far-blue I also want to use your fork, are you using this one?
> 
> https://github.com/far-blue/laminas-permissions-acl/tree/feature/php8.0_support

Yes :) Although I recommend you fork it and use your fork because if and when php 8 support is available in the mainline I'll switch over and delete my fork which might cause you some unexpected problems :)

**Comment by jmontoyaa** [comment]
Great! Thanks for the heads up @far-blue

**Comment by froschdesign** [comment]
@far-blue 
You are right, from the technical point of view you can install version 2.7 of laminas-servicemanager because of `^7.0`. Here I was completely wrong with the [caret version range](https://getcomposer.org/doc/articles/versions.md#caret-version-range-) which means for `^7.0`: 7.0 _and_ 7.1, 7.2, 7.3, 7.4.
I have hidden this completely for the PHP version. :man_facepalming:
I therefore apologise!

But another reason is relevant: version 2.7 of laminas-servicemanager is no longer supported. Therefore, with the support of version 8, this version was also removed in other components.
Please compare with:

* laminas-cache
* laminas-form
* laminas-feed
* laminas-inputfilter
* laminas-i18n
* …

**Comment by far-blue** [comment]
While I agree with your aim of encouraging people away from service manager v2 I think it would be unnecessarily cumbersome to force them to drop it until they move to php 8.

As service manager 2 doesn't support php 8 they will have to transition at some point but by allowing v2 to be supported in this package for the moment you are allowing them to update in stages, first by moving all their code to php8-supporting packages while staying on php7.4, and doing so in a controlled manner. Then they can move to php8.

As I've already found, there's enough juggling of packages to do when moving to php8 as it is because a lot of package developers are using the move as a chance to ditch older php versions and older dependency versions - causing me pain. I don't use service manager but I'd not want to inflict the same difficulties on other developers and hinder their move to php 8.

I would recommend not removing support for service manager v2 until you move to a version of permissions-all that doesn't support php7.4, thereby giving a migration path.

**Comment by PowerKiKi** [comment]
I agree with @far-blue. This PR is about adding support for PHP 8.0 and that goal is fulfilled as-is. I think the PR should be merged.

If/when dropping support for service-manager v2 become necessary it should be a separate PR.

**Comment by froschdesign** [comment]
@far-blue 
>  I would recommend not removing support for service manager v2 until you move to a version of permissions-all that doesn't support php7.4, thereby giving a migration path.

Unfortunately, you are overlooking an important point: If someone sets laminas-servicemanger explicitly to version 2 then version 2.7. or lower of laminas-permissions-acl is still used – nothing is changed. The next version 2.8 is then not installed, but everything works as before!

@PowerKiKi 
> If/when dropping support for service-manager v2 become necessary it should be a separate PR.

We have already dropped the support for this old and unsupported version in other components. This way we preserve the consistency.
So we are not doing anything new or deciding anything new here.

**Comment by far-blue** [comment]
Someone using permissions-acl 2.7 with service manager v2 should, in my opinion, be able to update to v2.8 of permissions-acl without having to move away from service manager v2. That puts them in the position where they are one step closer to php8 support before they actually move to php 8. As things stand with this PR, that is what will happen and it will make migration easier. I consider this a good thing.

If you have other, non-php8-related reasons for removing support for service manager v2 then this is not, in my opinion, the PR to do it in. Just raise another PR. Associating the removing of service manager v2 with this PR creates an implicit relationship which isn't there.

As I've covered above, I'd also urge you to not remove support for service manager v2 in this package until you actually stop supporting php7.x in this package, giving developers a chance to migrate their code. I'd also, at that point, suggest versioning this package a 3.x because, as @PowerKiKi points out, otherwise people may not notice they are being held back on updates due to blocking package dependencies.

I mean, feel free to release v2.8 and v3.0 at the same time with one of them supporting php7.x (and service manager 2.x) and the other supporting neither :) Then you get the best of both worlds - migration support for those who need to change code at the same time as a public statement of direction for the package.

But, again, not something for this PR.

**Comment by froschdesign** [comment]
> I'd also, at that point, suggest versioning this package a 3.x because, as @PowerKiKi points out, otherwise people may not notice they are being held back on updates due to blocking package dependencies.

Please see at semantic versioning: https://semver.org/#what-should-i-do-if-i-update-my-own-dependencies-without-changing-the-public-api

**Comment by far-blue** [comment]
> > I'd also, at that point, suggest versioning this package a 3.x because, as @PowerKiKi points out, otherwise people may not notice they are being held back on updates due to blocking package dependencies.
> 
> Please see at semantic versioning: https://semver.org/#what-should-i-do-if-i-update-my-own-dependencies-without-changing-the-public-api

Well, I was referring to dropping support for php7.x when suggesting a move to a 3.x version. That would be a backward-incompatible change, would it not? I mean, technically, if the code is actually php7.4 compatible then maybe you shouldn't change the version but if you intend on using php8 functionality then it could be be seen as preemptive. It was only a suggestion to make the lives of developers using the package easier. 

Either way, only an aside and irrelevant to this PR :)

**Comment by froschdesign** [comment]
@far-blue 
> Well, I was referring to dropping support for php7.x when suggesting a move to a 3.x version. That would be a backward-incompatible change, would it not?

If the public API is not changed, then no.

**Comment by PowerKiKi** [comment]
@froschdesign I created #16 to resolve the situation. Would you be willing to merge this PR and #16 immediately after, in order to finally get PHP 8.0 support ?

**Review by boesing — APPROVED** [review]
- Rebased against latest 2.8.x
- Dropped support for `laminas/laminas-servicemanager` v2
- Updated `composer.lock` (with PHP 7.3) to be in sync with `composer.json`

**Pull request merged**
